rfcs/text/0014_api_documentation.md

```diff
@@ -0,0 +1,356 @@
+  objects: ApiDeclaration[];
+  enums: ApiDeclaration[];
+  misc: ApiDeclaration[];
+  // We may add more here as we sit fit to pull out of `misc`.
+}
+```
+
+With this structure, the mdx files end up looking like:
+
+```
+### Start
+<DocDefinitionList data={[actionsJson.server.start]}/>
+### Functions
+<DocDefinitionList data={actionsJson.server.functions}/>
+### Interfaces
+<DocDefinitionList data={actionsJson.server.interfaces}/>
+```
+
+### PluginApi
+
+A plugins API is the component that is serialized into the json file. It is broken into public, server and common components.  `serviceFolders` is a way for the system to
+write separate mdx files depending on where each declaration is defined. This is because certain plugins (and core)
+are huge, and can't be rendered in a single page.
+
+
+```ts
+export interface PluginApi {
+  id: string;
+  serviceFolders?: readonly string[];
+  client: ScopeApi;
+  server: ScopeApi;
+  common: ScopeApi;
+}
+```
+
+## Manifest file changes
+
+**serviceFolders?: string[]**
+
+Used by the system to group services into sub-pages. Some plugins, like data and core, have such huge APIs they are very slow to contain in a single page, and they are less consummable by solution developers. The addition of an optional list of services folders will cause the system to automatically create a separate page with every API that is defined within that folder. The caveat is that core will need to define a manifest file in order to define its service folders...
+
+**Using a kibana.json file for core**
+
+For the purpose of API infrastructure, core is treated like any other plugin. This means it has to specify serviceFolders section inside a manifest file to be split into sub folders. There are other ways to tackle this - like a hard coded array just for the core folder, but I kept the logic as similar to the other plugins as possible.
+
+**teamOwner: string**
+
+Team owner can be determined via github CODEOWNERS file, but we want to encourage single team ownership per plugin. Requiring a team owner string in the manifest file will help with this and will allow the API doc system to manually add a section to every page that has a link to the team owner. Additional ideas are teamSlackChannel or teamEmail for further contact. 
+
+**summary: string**
+
+A brief description of the plugin can then be displayed in the automatically generated API documentation.
+
+## Technology: ts-morph vs api-extractor
+
+[Api-extractor](https://api-extractor.com/) is a utility built from microsoft that parses typescript code into json files that can then be used in a custom [api-documenter](https://api-extractor.com/pages/setup/generating_docs/) in order to build documentation. This is what we [have now](https://github.com/elastic/kibana/tree/master/docs/development), except we use the default api-documenter. Unfortunately, because our plugins aren’t separate packages, and the way api-extractor works, we can’t build cross plugin links this way.
```

Comment:
Since api-extractor-model supports loading several api.json files the API theoretically supports resolving cross-plugin links. We currently use kibana's top-level package.json file when running api-extractor so all api.json files use the "kibana" package name. If instead of supplying a `packageJsonFullPath` we supplied a generated `packageJson` with the name of the plugin we would get the correct package names in the generated api.json files https://github.com/elastic/kibana/blob/master/src/dev/run_check_published_api_changes.ts#L101

Reply:
I'll add that point to this section. I think I tried to explore this approach but wasn't able to get to it working. It's fully possible I overlooked something, or overestimated the effort/underestimated the feasibility. 

I still think it's worth trying out the current approach with ts-morph, and re-evaluating as time goes on.